### DISCUSSION TITLE
Compiler allows Nullable function to return Not Nullable value

### DISCUSSION THREAD

**Discussion by GuiAguirres** [dotnet/csharplang#8250] [Q&A]
When designing a function with a nullable return value, the compiler stays quiet about trying to return a not nullable value, as this:

```C#
public long? GetNullableValue()
{
    long value = 10;
    return value;
}
```

This function signature states a function that returns a Nullable<long>.

In my vision, the compiler should throw the same casting error when doing the opposite:

```C#
public long GetNullableValue()
{
    long? value = 10;
    return value;
}
```
Cannot implicitly convert type 'long?' to 'long'. An explicit conversion exists (are you missing a cast?)

# Motivation for this proposal:

The following function returns 0, instead of null:

``` C#
public long? GetNullableValue()
{
    long[] values = [10, 5, 2];
    return values.Where(x => x == 7).LastOrDefault();
}
```

But this one returns null:

```C#
public long? GetNullableValue()
{
    long[] values = [10, 5, 2];
    return values.Where(x => x == 7).Select(x => (long?)x).LastOrDefault();
}
```

Environment: macOS Air M2
dotnet 8

**Comment by CyrusNajmabadi** [comment]
This is by design.  All types are implicitly convertible to the nullable version of that type.  

> When designing a function with a nullable return value, the compiler stays quiet about trying to return a not nullable value, as this:

Correct.  The `long` value 10 is perfectly representable as a `long?` so the compiler allows implicitly converting this.  Just like it allows implicitly converting a `string` to an `object`, and many other implicit conversions. 

> Motivation for this proposal:

You've used `LastOrDefault` on an `IEnumerable<long>` and an `IEnumerable<long?>` respectively.  The `default(long)` is just `0`, while `default(long?)` is null.

**Comment by TahirAhmadov** [comment]
If your method has a nullable value type return type, and all returns have a known not-null value, it is indeed a potential source of bugs. You can write an analyzer which would warn in those situations.

For `IEnumerable<T>` where `T : struct`, create and use an extension method like `LastOrNull` which returns `T?`, instead of `LastOrDefault`. Then `longs.LastOrNull()` will be `long?`, as desired.
